src/gridcoin/tally.cpp

```diff
@@ -290,6 +292,69 @@ class ResearcherTally
         }
     }
 
+    //!
+    //! \brief Record a block's MRC research reward data in the tally
+    //! \param pindex: Contains information about the block to record
+    //!
+    void RecordMRCRewardBlock(const CBlockIndex* const pindex)
+    {
+        for (const auto& mrc_researcher : pindex->m_mrc_researchers) {
+            Cpid cpid = mrc_researcher->m_cpid;
+
+            ResearchAccount& account = m_researchers[cpid];
+
+            account.m_total_research_subsidy += mrc_researcher->m_research_subsidy;
+
+            // MRC's are paid on the last block prior to the staked block (i.e. pprev), but recorded in the pindex
+            // where the claim is. This is a little tricky.
+            //
+            // TODO: This probably doesn't work correctly given the implicit cast and should be removed. It isn't
+            // used in accrual calculations, only reporting.
+            if (mrc_researcher->m_magnitude > 0) {
+                account.m_accuracy++;
+                account.m_total_magnitude += pindex->pprev->Magnitude();
+            }
+
+            if (account.m_first_block_ptr == nullptr) {
+                account.m_first_block_ptr = pindex->pprev;
+                account.m_last_block_ptr = pindex->pprev;
+            } else {
+                assert(pindex->pprev->nHeight > account.m_last_block_ptr->nHeight);
+                account.m_last_block_ptr = pindex->pprev;
+            }
+        }
+    }
+
+    //!
+    //! \brief Finds the last reward block prior to the reward that is being disassociated from the tally.
+    //!
+    //! \param cpid: The CPID of the research account to search for.
+    //! \param pindex: The starting point of the search (pindex->pprev is actually used).
+    //! \return
+    //!
+    const CBlockIndex* FindLastRewardBlock(const Cpid cpid, const CBlockIndex* pindex)
+    {
+        // We have to find the last block pointer for the reward not including this one that is being rolled back.
+        // The search here is more complicated now that we have mrc, because the previous reward could be either a stake or
+        // mrc, so we have to look for both.
+        pindex = pindex->pprev;
+
+        while (pindex)
+        {
+            // If it is a stake with this cpid, then stop
+            if (pindex->ResearchSubsidy() > 0 && pindex->GetMiningId() == cpid) break;
+
+            // If it is an MRC with this cpid and non-zero reward, then stop
+            for (const auto& mrc_researcher : pindex->m_mrc_researchers) {
+                if (mrc_researcher->m_cpid == cpid && mrc_researcher->m_research_subsidy > 0) break;
```

Comment:
```suggestion
                if (mrc_researcher->m_cpid == cpid && mrc_researcher->m_research_subsidy > 0) {
                    pindex = pindex->pprev;
                    break;
                }
```
The true last reward would be the block before the MRC claim. See the comments below for why this is necessary.

Reply:
I think you are right. This is tricky stuff, because the accrual account last block/first block has to record the block at which the accrual ends but the actual block that the mrc is recorded is the block with the claim/coinstake of the mrc. In the case of normal rewards this is coincident, but with mrc, the accrual block context for the reward is always one block behind. I fixed the other parts of the tally to deal with this, but I think missed it in this one place. I think this commit fixes all of the below too.